README.md
# webmention-cli

**A command-line interface for Webmention written in Ruby.**

[![Gem](https://img.shields.io/gem/v/webmention-cli.svg?logo=rubygems&style=for-the-badge)](https://rubygems.org/gems/webmention-cli)
[![Downloads](https://img.shields.io/gem/dt/webmention-cli.svg?logo=rubygems&style=for-the-badge)](https://rubygems.org/gems/webmention-cli)
[![Build](https://img.shields.io/github/actions/workflow/status/jgarber623/webmention-cli/ci.yml?branch=main&logo=github&style=for-the-badge)](https://github.com/jgarber623/webmention-cli/actions/workflows/ci.yml)

## Installation

Before installing and using webmention-cli, you'll want to have [Ruby](https://www.ruby-lang.org) 2.7 (or newer) installed.

```sh
$ gem install webmention-cli
```

## Usage

webmention-cli makes available the following commands:

### `webmention endpoint <target>`

Discover the Webmention endpoint for a `<target>` URL.

```sh
$ webmention endpoint https://sixtwothree.org
https://sixtwothree.org/webmentions
```

The command returns a non-zero exit code if the `<target>` URL does not advertise a Webmention endpoint (or if an HTTP error was encountered).

### `webmention send <source> <target>`

Send a webmention from a `<source>` URL to a `<target>` URL.

```sh
$ webmention send https://sixtwothree.org/posts/an-engineer-walks-into-a-design-sprint https://adactio.com/journal/6246
202 Accepted
```

The command will return either an HTTP status _or_ the value of the response's `Location` header (if provided _and_ the response's HTTP status is `201 Created`).

A non-zero exit code is returned if the HTTP status is _not_ within the 200 range (or if an HTTP error was encountered).

#### Send a webmention with a vouch URL

```sh
$ webmention send https://sixtwothree.org/posts/now-accepting-webmentions https://aaronparecki.com --vouch https://adactio.com/links/9229
202 Accepted
```

The command with a vouch URL included responds in the same fashion as noted above.

### `webmention verify <source> <target>`

Confirm whether or not a `<source>` URL links to a `<target>` URL.

```sh
$ webmention verify https://kartikprabhu.com/notes/re-launching-franciscms https://sixtwothree.org/posts/launching-franciscms-onto-the-indieweb
```

The command will return a zero exit code if `<source>` links to `<target>` and a non-zero exit code in all other cases.

#### Verify a webmention with a vouch URL

```sh
$ webmention verify https://kartikprabhu.com/notes/re-launching-franciscms https://sixtwothree.org/posts/launching-franciscms-onto-the-indieweb --vouch https://adactio.com/links/9229
```

The command will return a zero exit code if `<source>` links to `<target>` _and_ the provided vouch URL links to the `<source>`'s domain. A non-zero exist code is returned in all other cases.

## Acknowledgments

webmention-cli wouldn't exist without Webmention and the hard work put in by everyone involved in the [IndieWeb](https://indieweb.org) movement.

webmention-cli is written and maintained by [Jason Garber](https://sixtwothree.org).

## License

webmention-cli is freely available under the [MIT License](https://opensource.org/licenses/MIT). Use it, learn from it, fork it, improve it, change it, tailor it to your needs.


CHANGELOG.md
# Changelog

## 3.0.1 / 2023-12-30

- Add `source_code_uri` to metadata (caf4022)

## 3.0.0 / 2023-12-12

- **Breaking change:** Update Thor and Webmention dependencies (#90) (b89a756)
- Address RuboCop warnings (#89) (9a405bf)
- Code cleanup (#88) (445984f)
- **Breaking change:** Update Ruby versions (#87) (8a36cd3)
  - Development: 2.7.8
  - Minimum supported: 2.7

## 2.0.0 / 2022-05-20

- **Breaking change:** Rewrite gem code (35e9442)
- **Breaking change:** Update minimum supported Ruby version to 2.6 (b8bca32)
- Update maximum required Ruby version to < 4 (d5ca374)
- Update development Ruby version to 2.6.10 (cd27f96)
- Remove Reek (95096d0)

## 1.5.0 / 2020-08-23

- Update indieweb-endpoints and webmention dependencies (209a75b)

## 1.4.0 / 2020-08-23

- Update indieweb-endpoints and webmention dependencies (32bd719)

## 1.3.0 / 2020-05-26

- Update webmention dependency (4540771)

## 1.2.0 / 2020-05-18

- Update indieweb-endpoints, webmention, and webmention-verification gems (2ad2f78)
- Move development dependencies to Gemfile per current Bundler conventions (434f785)
- Update development Ruby version to 2.5.8 (1438a62)

## 1.1.0 / 2020-04-17

- Update webmention dependency (21695df)

## 1.0.0 / 2020-02-08

- Rewrite gem code
- Rewrite specs
- Update development and runtime dependencies to latest versions (67744c1)
- Change supported Ruby versions to >= 2.5 and < 2.8 (1b30b8c)
- Update project Ruby version to 2.5.7 (21e94b6)

## 0.1.1 / 2019-01-03

- Expand supported Ruby versions to include 2.6 (8930501)

## 0.1.0 / 2018-08-14

- Initial release!


lib/webmention/cli/runner.rb
# frozen_string_literal: true

module Webmention
  module CLI
    class Runner < Thor
      include Thor::Actions

      package_name "Webmention"

      map ["-v", "--version"] => :version

      # :nocov:
      def self.exit_on_failure?
        true
      end
      # :nocov:

      desc "endpoint <target>", "Discover the Webmention endpoint for <target> URL"
      def endpoint(target)
        url = Webmention::Url.new(target)

        exit 1 unless url.webmention_endpoint?

        say url.webmention_endpoint
      end

      desc "send <source> <target>", "Send a webmention from <source> URL to <target> URL"
      option :vouch, desc: "Submit a <vouch> URL"
      def send(source, target)
        response = Webmention.send_webmention(source, target, vouch: options[:vouch])

        exit 1 unless response.ok?

        code = response.code
        location = response.headers["Location"]

        say(code == 201 && location ? location : "#{code} #{response.reason}")

        exit 1 unless code.between?(200, 299)
      end

      desc "verify <source> <target>", "Verify that <source> URL links to <target> URL"
      option :vouch, desc: "Verify that <vouch> URL mentions the <source> URL domain"
      def verify(source, target)
        verification = Webmention.verify_webmention(source, target, vouch: options[:vouch])

        exit 1 unless verification.verified?
      end

      desc "version", "Print version information"
      def version
        say "webmention-cli version #{Webmention::CLI::VERSION}"
      end
    end
  end
end
